sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -306,6 +306,47 @@ object OptimizeIn extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * SimplifyJoinCondition:
+ * Changes the join condition of type (key1=key2) || (isnull(key1)&&isnull(key2))
+ * to key1 <=> key2 to prevent join from converting to BNLJ in physical planning
+ * which is costly implementation of join.
+ */
+object SimplifyJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsPattern(JOIN), ruleId) {
+    case join@Join(_, _, _, Some(joinCondition), _) =>
```

Comment:
I think it's beneficial for `Filter` as well?

Reply:
Could we update `SimplifyConditionalsInPredicate` to optimize this condition?
```scala
    case Or(EqualTo(l, r), And(IsNull(c1), IsNull(c2)))
      if (l.semanticEquals(c1) && r.semanticEquals(c2)) ||
        (l.semanticEquals(c2) && r.semanticEquals(c1)) =>
      EqualNullSafe(simplifyConditional(l), simplifyConditional(r))
```

Reply:
Sure. SimplifyConditionalsInPredicate will work when Case When or If conditional expression is present in tree.